Added support for Common Type Attributes

shubhamnarlawar77 opened the pull request.

shubhamnarlawar77 committed:
Added support for Common Type Attributes

shubhamnarlawar77 commented:
@marxin : review request

shubhamnarlawar77 commented:
The purpose of these 2 branches `csmith-gcc-v2-generalized` and `csmith-gcc-v2-generalized-review` is to cleanup and generalize base branch `csmith-gcc-v2`

shubhamnarlawar77 commented:
Please check below code in `Function.cpp`

`void
FunctionAttribute::OutputAttribute(std::ostream &out, string option)
{
        if(!attr_emitted){
                out << " __attribute__((" << attribute << option;
                attr_emitted = true;
        }
        else
                out << ", " << attribute << option;
}`
I want to shift this checking of `attr_emitted` to `Function::OutputForwardDecl()` so that I can generalize `Common Function Attributes` for `Type Attributes` also.

Now, generation of type attributes is messed due to `attr_emitted` and also function & types generating randomnly. Please generate a random program using `--function-attributes` to check it.

shubhamnarlawar77 commented:
Did you check above comment? Should I push revised patch?

Note - The above patch is pushed to get things clear for generalizing attributes.

shubhamnarlawar77 commented:
Did you check above comment? Should I push revised patch?

Note - The above patch is pushed to get things clear for generalizing attributes.

marxin commented on src/Type.cpp:
```diff
@@ -71,7 +72,20 @@ Type *Type::void_type = NULL;
 //////////////////////////////////////////////////////////////////////
+
+void
+GenerateAttributes()
```
The function name is misleading, it's actually not generating any attributes.

shubhamnarlawar77 replied:
Yeah. It is because I have not added other attributes like `aligned, visibility....`

marxin replied:
It does not clarify the function name.

marxin commented on src/Type.cpp:
```diff
@@ -71,7 +72,20 @@ Type *Type::void_type = NULL;
+void
+GenerateAttributes()
+{
+		vector<string> common_type_attributes = {"deprecated", "unused", "transparent_union"};
```
Please do not use a for cycle, use sequence of `push_back`, it's more readable.

marxin commented on src/Type.cpp:
```diff
@@ -317,6 +331,8 @@ Type::Type(vector<const Type*>& struct_fields, bool isStruct, bool packed,
     else
         eType = eUnion;
     sid =  sequence++;
+	struct_attr_flag = false;
```
You should not need this.

shubhamnarlawar77 replied:
I need this to identify whether given type is struct or not. One should not generate type attributes to other variables.

marxin replied:
Can you use existing fields to distinguish `struct` and `enum`?

marxin commented on src/Type.cpp:
```diff
@@ -71,7 +72,20 @@ Type *Type::void_type = NULL;
+		vector<string>::iterator itr;
+		type_attributes.push_back(new MultiValuedFunctionAttribute("visibility", FuncAttrProb, {"default", "hidden", "protected", "internal"}));
+		for(itr = common_type_attributes.begin(); itr < common_type_attributes.end(); itr++)
+			type_attributes.push_back(new BooleanFunctionAttribute(*itr, FuncAttrProb));
```
Are you sure about `FuncAttrProb`?

shubhamnarlawar77 replied:
My mistake. Need to change it to `TypeAttrProb`

marxin commented on src/Type.cpp:
```diff
@@ -54,6 +54,7 @@
 #include "Probabilities.h"
 #include "DepthSpec.h"
 #include "Enumerator.h"
+#include "Function.h"
```
Does not makes sense to include a `Function.h` here.

shubhamnarlawar77 replied:
It is required to include infrastructure defined in `Function.h` and even if I moved to a separate file I need to include whereever I want to use that infrastructure.

marxin replied:
Yep, please read my Review message, it's suggesting such a refactoring.

marxin reviewed (changes requested):
Ok, I would like to ask you first to rebase the original branch which we have and rename all `BooleanFunctionAttribute` so that they don't contain 'Function'. You would need to put them into a separate file probably.

shubhamnarlawar77 commented:
I have pushed suggested changes. Please check it.

1. This is demo branch to get things sort out faster and I will rebase original branch later.
2. Further, I will print attributes in list like `__attribute__((hot, cold, used, unused..))`. I have undo it due to conflicting use of `attr_emitted` by `Function` and `Type` class.
3. At last I will try to move infrastructure to new file.

marxin commented:
> 1. This is demo branch to get things sort out faster and I will rebase original branch later.

Yes, the current version looks reasonable fine for me!

> 2. Further, I will print attributes in list like `__attribute__((hot, cold, used, unused..))`. I have undo it due to conflicting use of `attr_emitted` by `Function` and `Type` class.

Can you use attributes likes `hot` for types or variables?

> 3. At last I will try to move infrastructure to new file.

Yes, that can be next step to rebase all what we currently have + add the type attributes.
You're doing well!

shubhamnarlawar77 commented:
> > 1. This is demo branch to get things sort out faster and I will rebase original branch later.
> 
> Yes, the current version looks reasonable fine for me!
> 
> > 1. Further, I will print attributes in list like `__attribute__((hot, cold, used, unused..))`. I have undo it due to conflicting use of `attr_emitted` by `Function` and `Type` class.
> 
> Can you use attributes likes `hot` for types or variables?
>
Sure. 
I want to ask a edit. Currently, it prints each attribute as single `__attribute__((hot))` `__attribute__((cold))`. Can I remove probability option from base class and use probability in output phase to print all attributes in single `__attribute__((.......))`?

marxin commented:
> I want to ask a edit. Currently, it prints each attribute as single `__attribute__((hot))` `__attribute__((cold))`. Can I remove probability option from base class and use probability in output phase to print all attributes in single `__attribute__((.......))`?

Ok, so this topic is still open. I would recommend to come up with a class (e.g. `AttributeGenerator`), that will have `vector<FunctionAttribute*> attributes` and also a flag (called `attribute_emitted`). Then the class will do:
```
        for(itr = attributes.begin(); itr < attributes.end(); itr++)
                (*itr)->OutputAttributes(out);
```

and later then eventually print `))` if needed (if an attribute is emitted)`. That would a handy encapsulation class which you'll use for functions, types, variables, ...

shubhamnarlawar77 commented:
> > I want to ask a edit. Currently, it prints each attribute as single `__attribute__((hot))` `__attribute__((cold))`. Can I remove probability option from base class and use probability in output phase to print all attributes in single `__attribute__((.......))`?
> 
> Ok, so this topic is still open. I would recommend to come up with a class (e.g. `AttributeGenerator`), that will have `vector<FunctionAttribute*> attributes` and also a flag (called `attribute_emitted`). Then the class will do:
> 
> ```
>         for(itr = attributes.begin(); itr < attributes.end(); itr++)
>                 (*itr)->OutputAttributes(out);
> ```
> 
> and later then eventually print `))` if needed (if an attribute is emitted)`. That would a handy encapsulation class which you'll use for functions, types, variables, ...

Done.
1. Are names of object of AttributeGenerator fine to you?
      For functions - func_attr_generator
      for types - type_attr_generator
2. I have added an alignment factor in Alignment class as function can take alignement upto 2^16 where as types 2^8. (IIRC once you said in a mail).
3. When I add hot attribute types, I get a warning as `Hot attribute no applicable to types`. Hence, I have not added `hot` attribute for types.

shubhamnarlawar77 commented:
Is patch ready for rebasing base branch?

marxin commented:
By rebasing I mean modifying existing commits so that we won't introduce a class (e.g. `MultiValuedFunctionAttribute`), that we will then rename.

shubhamnarlawar77 commented:
Yes. I mean editing those commits.
Is the current state fine?

marxin commented:
I have some comments about it, but yes. Please rebase it now and then we can enhance the rebased branch.

shubhamnarlawar77 commented:
I have rebased base branch `csmith-gcc-v2`. I have added attribute infrastructure in separate files named `Attribute.cpp Attribute.h`. Please check and suggest changes if any.

marxin commented:
It looks pretty fine. I'm thinking why is `always_inline` not a boolean attribute?

marxin commented:
Please make a new pull request and close all the existing.

shubhamnarlawar77 commented:
> It looks pretty fine. I'm thinking why is `always_inline` not a boolean attribute?

Merging `always_inline` into infrastructure causes Csmith to generate non-compilable programs. So, we agreed to look into it later. Maybe while upstreaming patches with the community, we can discuss with them. Let us keep it as a separate commit for now if it is fine to you?

marxin commented:
> Merging `always_inline` into infrastructure causes Csmith to generate non-compilable programs. So, we agreed to look into it later. Maybe while upstreaming patches with the community, we can discuss with them. Let us keep it as a separate commit for now if it is fine to you?

ok

shubhamnarlawar77 commented:
> Please make a new pull request and close all the existing.

Sure. But I am currently rebasing base branch. From which branch I am supposed to sent pull request. 
Can we discuss on the commit itself?

shubhamnarlawar77 commented:
Should I cherry-pick remaining function attributes commits now?
Later, I will create pull request for type attributes.

marxin commented:
> Should I cherry-pick remaining function attributes commits now?
Yes please, I'll review it all.

shubhamnarlawar77 commented:
> > Should I cherry-pick remaining function attributes commits now?
> > Yes please, I'll review it all.

I have edited `aligned` commit using cherry-pick in base branch. Please check it so that I can cherry-pick remaining function attribute commits.

marxin commented:
> I have edited `aligned` commit using cherry-pick in base branch. Please check it so that I can cherry-pick remaining function attribute commits.

As mentioned I want to see a new pull request that will contain following commits:
```
9576a9c (shub/csmith-gcc-v2) Added support for Aligned Function Attribute
49a0b6c Added support for Common Function Attributes
e328443 Added support for Function attribute always_inline
```

shubhamnarlawar77 commented:
> > I have edited `aligned` commit using cherry-pick in base branch. Please check it so that I can cherry-pick remaining function attribute commits.
> 
> As mentioned I want to see a new pull request that will contain following commits:
> 
> ```
> 9576a9c (shub/csmith-gcc-v2) Added support for Aligned Function Attribute
> 49a0b6c Added support for Common Function Attributes
> e328443 Added support for Function attribute always_inline
> ```

Except this right? `e328443 Added support for Function attribute always_inline`

marxin commented:
> Except this right? `e328443 Added support for Function attribute always_inline`

yes

shubhamnarlawar77 commented:
> > Except this right? `e328443 Added support for Function attribute always_inline`
> 
> yes

Is it fine now?

marxin commented:
> Is it fine now?

No, this pull request contains just a single commit.

shubhamnarlawar77 commented:
> > Is it fine now?
> 
> No, this pull request contains just a single commit.

Done. Please check the pull request. Is it the way you wanted a pull request containing `Common Function Attributes` and `Aligned Function Attribute`?

marxin commented:
Please close this.
